specifyweb/specify/tree_extras.py

```diff
@@ -633,96 +639,73 @@ def print_paths(table, depth):
-        },
-            "child" : {
-              f"{table.capitalize()} ID" : childID,
-              "Full Name" : childName,
-              "Bad Rank" : childRank
-        }} for parentID, parentName, childID, childName, childRank in results],
-        "localizationKey" : "badTreeStructureInvalidRanks",
+                "parent": {
+                    f"{table.capitalize()} ID": result.parent.id,
+                    "Full Name": result.parent.fullname
+                },
+                "child": {
+                    f"{table.capitalize()} ID": result.id,
+                    "Full Name": result.fullname,
+                    "Bad Rank": result.rankid
+                }
+            } for result in results
+        ],
+        "localizationKey": "badTreeStructureInvalidRanks",
     }
-    bad_ranks_count = cursor.rowcount
+    bad_ranks_count = results.count()
     formattedResults["badRanks"] = bad_ranks_count
-    if bad_ranks_count > 0 : raise AssertionError(
-        f"Bad Tree Structure: Found {bad_ranks_count} case(s) where node rank is not greater than its parent",
-        formattedResults)
+    if bad_ranks_count > 0:
+        raise AssertionError(
+            f"Bad Tree Structure: Found {bad_ranks_count} case(s) where node rank is not greater than its parent",
+            formattedResults
+        )
 
     # Get the tree ranks in leaf -> root order.
-    cursor.execute("select distinct rankid from {} order by rankid desc".format(table))
-    ranks = [rank for (rank,) in cursor.fetchall()]
+    ranks = tree_model.objects.values_list('rankid', flat=True).distinct().order_by('-rankid')
     depth = len(ranks)
 
     # Construct a path enumeration for each node and set the
-    # nodenumbers according to the lexical ordering of the paths. This
-    # ensures ancestor node numbers come before descendents and
-    # sibling nodes get consecutive ranges.
-    cursor.execute("set @rn := 0")
-    cursor.execute((
-        "update {table} t\n"
-        "join (select @rn := @rn + 1 as nn, p.id as id from \n"
-        "         (select t0.{table}id as id, {path} as path\n"
-        "          from {table} t0\n"
-        "          {parent_joins}\n"
-        "          order by path) p\n"
-        ") r on t.{table}id = r.id\n"
-        # Set the highestchildnodenumber to be the same as the
-        # nodenumber.  This is correct for leaves, and interior nodes
-        # will be adjusted in the next step.
-        "set t.nodenumber = r.nn, t.highestchildnodenumber = r.nn\n"
-    ).format(
-        table=table,
-        path=path_expr(table, depth),
-        parent_joins=parent_joins(table, depth),
-    ))
+    # nodenumbers according to the lexical ordering of the paths.
+    nodes = tree_model.objects.all().order_by('path')
+    with transaction.atomic():
+        for i, node in enumerate(nodes, start=1):
+            node.nodenumber = i
+            node.highestchildnodenumber = i
+            node.save()
```

Comment:
Sorry for the stray comment, but I'd highly recommending not this approach (and, instead, using raw sql). Also, the previous approach renumbers the entire tree -- so is this (restricting to a specific tree, based on commit message) done for performance reasons? If so, then that's technically not perfect. Actually, this (restricting to a specific tree) will only be correct when you are doing moves (which keeps total count same)

Anything else, merge, add, delete changes the node numbers of that tree (either increments or decrements). In doing so, the rest of the trees need to be updated. Well, technically, you can just update the trees with root node numbers after the current tree in case of inserts, and the ones before for deletes, for merges, it's complicated. I don't think complexity is worth here, given that you guys would want to remove the node numbers entirely, so just keep it be the previous code.